src/common/datapatterns.c

```diff
@@ -388,22 +398,32 @@ xdd_datapattern_get_datap_from_offset(worker_data_t *wdp) {
 		 *
 		 * In the unaligned case we will allocate the IO buffer for the worker and copy
 		 * the correct data to it.
-		 */  
-		offset = wdp->wd_task.task_byte_offset % tdp->td_dpp->data_pattern_length;
-		if ((tdp->td_dpp->data_pattern_length - offset) < (size_t)tdp->td_xfer_size) { 
-			// The target's xfer_size does not evenly divide into the whole file's
-			// size, so we have to allocate a buffer (if not already allocated) to
-			// hold the data an copy it over.
-			if (!wdp->wd_bufp_allocated) {
-				buff = xdd_init_io_buffers(wdp);
-				wdp->wd_task.task_datap = buff;
-			} else {
-				buff = wdp->wd_task.task_datap;
-			}
-			memcpy_size = tdp->td_dpp->data_pattern_length - offset; 
-			memcpy(buff, &(tdp->td_dpp->data_pattern[offset]), memcpy_size);
-			memcpy(&buff[memcpy_size], tdp->td_dpp->data_pattern, (tdp->td_xfer_size - memcpy_size));
-		} else {
+		 */
+
+		// Calculate the offset within the datapattern buffer
+		offset = wdp->wd_task.task_byte_offset % buffer_size;
+		// Set the remaining size to the transfer size
+		remaining_size = tdp->td_xfer_size;
+
+		// If the buffer has already allocated, we can copy the data
+		// into buffer. We also handle the offset to ensure the data
+		// is being copied correctly
+		if (wdp->wd_bufp_allocated) {
+			buff = wdp->wd_task.task_datap;
+			unsigned char *buff_ptr = buff;
+			xdd_datapattern_copy_data_to_buffer(buff_ptr, tdp->td_dpp->data_pattern, buffer_size, offset, remaining_size);
+		}
+		// The target's xfer_size does not evenly divide into the whole file's
+		// size, so we have to allocate a buffer (if not already allocated) to
+		// hold the data an copy it over.
+		else if ((buffer_size - offset) < remaining_size) {
```

Comment:
Move this `else if` up to the `}` above. Then move the comment explaining the `else if` into the `else if` statement. Follow the coding style of the rest of XDD.